### PR TITLE
New local starting server

### PR DESCRIPTION
**Pull request by ygoumaz**
This new simple server allows to start Webots from a remote machine. The world and startup parameters are given in a shared folder. 

This server is added for macOS compatibility of `webots_ros2`, so that the Docker container can start Webots on the host machine, but this server could also be used in other situations.

### CODE REVIEW

File: local_simulation_server.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import subprocess
+import sys
+from os import walk
+
+HOST = ""  # Any host can connect
+PORT = int(sys.argv[1])  # Port to listen on
+
+shared_folder = None
+webots = None
+
```
Comment (omichel) [inline local_simulation_server.py]: PEP8 compliance (flake8):
```suggestion


```

File: local_simulation_server.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+def start_webots(connection):
+    #shared_folder_str = shared_folder.decode("utf-8")
```
Comment (omichel) [inline local_simulation_server.py]: Can we remove this line?
```suggestion
```

File: local_simulation_server.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    filenames = next(walk(shared_folder), (None, None, []))[2]
+    worlds = list(filter(lambda file: file.endswith('.wbt'), filenames))
+
+    if(len(worlds) == 0):
+        return -1
+    if(len(worlds) > 1):
+        return -2
+    world_file = os.path.join(shared_folder, worlds[0])
+
+    arguments = []
```
Comment (omichel) [inline local_simulation_server.py]: Unused:
```suggestion
```

File: local_simulation_server.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    connection.sendall(b"ACK")
+    subprocess.call(['/usr/bin/open', '-W', '-n', '-a', '/Applications/Webots.app', '--args', world_file, *lines])
+
+    os.remove(world_file)
+    if 'launch_args.txt' in filenames:
+        os.remove(os.path.join(shared_folder, 'launch_args.txt'))
+
+    return 1
+
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion


```

File: local_simulation_server.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+def keyboardInterruptHandler(signal, frame):
+    if(shared_folder):
+        if(os.path.isdir(shared_folder)):
+            for filename in os.listdir(shared_folder):
+                filepath = os.path.join(shared_folder, filename)
+                if os.path.isfile(filepath):
+                    os.remove(filepath)
+    exit(0)
+
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion


```

File: local_simulation_server.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    s.bind((HOST, PORT))
+    s.listen()
+    conn, addr = s.accept()
+    with conn:
+        print(f"Connected by {addr}")
+        while True:
+            data = conn.recv(1024)
+            if not data:
+                break
+            shared_folder = data.decode("utf-8")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
            shared_folder = data.decode('utf-8')
```

File: local_simulation_server.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                    os.remove(filepath)
+    exit(0)
+
+signal.signal(signal.SIGINT, keyboardInterruptHandler)
+with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+    s.bind((HOST, PORT))
+    s.listen()
+    conn, addr = s.accept()
+    with conn:
+        print(f"Connected by {addr}")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
        print(f'Connection from {addr}')
```

File: local_simulation_server.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        while True:
+            data = conn.recv(1024)
+            if not data:
+                break
+            shared_folder = data.decode("utf-8")
+            success = start_webots(conn) if(os.path.isdir(shared_folder)) else 0
+            if success == 0:
+                conn.sendall(b"FAIL0")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
                conn.sendall(b'FAIL0')
```

File: local_simulation_server.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                print(f"The shared folder '{data_str}' doesn't exist.")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
                print(f"The shared folder '{shared_folder}' doesn't exist.", file=sys.stderr)
```

File: local_simulation_server.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+            elif success == -1:
+                conn.sendall(b"FAIL1")
+                print(f"No world could be found in the shared folder.")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
                print('No world could be found in the shared folder.' file=sys.stderr)
```

File: local_simulation_server.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+            elif success == -2:
+                conn.sendall(b"FAIL2")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
                conn.sendall(b'FAIL2')
```

File: local_simulation_server.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+                print(f"More than one world was found in the shared folder.")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
                print('More than one world was found in the shared folder.' file=sys.stderr)
```

File: local_simulation_server.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+            elif success == 1:
+                print(f"Webots was executed successfully.")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
                print('Webots was executed successfully.')
```

File: local_simulation_server.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
+import os
+import signal
+import socket
+import subprocess
+import sys
+from os import walk
+
+HOST = ""  # Any host can connect
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
HOST = ''  # Any host can connect
```

File: local_simulation_server.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+    arguments = []
+    lines = []
+    if 'launch_args.txt' in filenames:
+        launch_args_file = open(os.path.join(shared_folder, 'launch_args.txt'), 'r')
+        lines = launch_args_file.readlines()
+        for i in range(len(lines)):
+            lines[i] = lines[i].strip()
+
+    connection.sendall(b"ACK")
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
    connection.sendall(b'ACK')
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Comment by ygoumaz** [comment]
Sorry for the high number of unused elements and mistakes in the code. I think an iteration of cleanup that I made was not committed and lost somewhere.

File: local_simulation_server.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
+import os
```
Comment (omichel) [inline local_simulation_server.py:19]: We should add a header similar to the other python scripts here with the license and description.

File: local_simulation_server.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+shared_folder = None
+webots = None
+
+
+def start_webots(connection):
+    filenames = next(walk(shared_folder), (None, None, []))[2]
+    worlds = list(filter(lambda file: file.endswith('.wbt'), filenames))
+
+    if(len(worlds) == 0):
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
    if len(worlds) == 0:
```

File: local_simulation_server.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+        return -1
+    if(len(worlds) > 1):
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
    if len(worlds) > 1:
```

File: local_simulation_server.py [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+    os.remove(world_file)
+    if 'launch_args.txt' in filenames:
+        os.remove(os.path.join(shared_folder, 'launch_args.txt'))
+
+    return 1
+
+
+def keyboardInterruptHandler(signal, frame):
+    if(shared_folder):
+        if(os.path.isdir(shared_folder)):
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
    if shared_folder:
        if os.path.isdir(shared_folder):
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Comment by ygoumaz** [comment]
The server is now simpler than before. It doesn't clean the shared folder anymore. Also, it is able to detect if the ROS process died and sends a message to the client once Webots is closed.

**Review by omichel — COMMENTED** [review]
It looks good to me.

However, as I understand, the user has to launch this local server repeatedly as he tries different things on the ROS side. I believe it would be more convenient for the user if the server would be able to accept a new connection a previous connection has closed, e.g., put everything inside a loop that will loop when the client closes the connection and that could be broken with <kbd>Ctr</kbd> + <kbd>C</kbd>.

**Comment by ygoumaz** [comment]
The server doesn't shutdown anymore when ROS or Webots exit. Instead we continue to listen for a new session.

In addition, Webots is started with the world and arguments given in the tcp message by the ROS client. I also added the possibility to define WEBOTS_HOME as target installation to run Webots.

File: local_simulation_server.py [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+        if 'WEBOTS_HOME' in os.environ:
+            cmd[0] = os.path.join(os.environ['WEBOTS_HOME'], 'webots')
+        else:
+            print('WEBOTS_HOME is not defined. Falling back to default installation folder `/Applications/Webots.app`')
+            cmd[0] = '/Applications/Webots.app/Contents/MacOS/webots'
+
+    try:
+        webots_process = subprocess.Popen(cmd)
+    except FileNotFoundError:
+        message = 'FAIL: Webots could not be found on the host.'
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
        message = f'FAIL: \'{cmd[0]}\'could not be found on the host.'
```

File: local_simulation_server.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+tcp_socket.listen()
+while True:
+    print(f'Waiting for connection on port {PORT}...')
+    connection, address = tcp_socket.accept()
+
+    print(f'Connection from {address}')
+    data = connection.recv(1024)
+    cmd = data.decode('utf-8').split(' ')
+
+    world_file = cmd[1]
```
Comment (omichel) [inline local_simulation_server.py]: The world file is not necessarily the first argument, it is usually specified after the options.

File: local_simulation_server.py [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+    try:
+        test_webots_cmd = subprocess.Popen([cmd[0]])
+    except FileNotFoundError:
+        print('`webots` command could not be found. Falling back to WEBOTS_HOME environment variable.')
+        if 'WEBOTS_HOME' in os.environ:
+            cmd[0] = os.path.join(os.environ['WEBOTS_HOME'], 'webots')
+        else:
+            print('WEBOTS_HOME is not defined. Falling back to default installation folder `/Applications/Webots.app`')
+            cmd[0] = '/Applications/Webots.app/Contents/MacOS/webots'
+
```
Comment (omichel) [inline local_simulation_server.py]: I would not rely on `WEBOTS_HOME` to make things simpler here and not specific to macOS:
```suggestion
```

File: local_simulation_server.py [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+tcp_socket.bind((HOST, PORT))
+tcp_socket.listen()
+while True:
+    print(f'Waiting for connection on port {PORT}...')
+    connection, address = tcp_socket.accept()
+
+    print(f'Connection from {address}')
+    data = connection.recv(1024)
+    cmd = data.decode('utf-8').split(' ')
+
```
Comment (omichel) [inline local_simulation_server.py]: We should check that `cmd[0].endswith('webots')` to ensure somebody doesn't send a `rm -rf /` command from the outside.

**Review by omichel — CHANGES_REQUESTED** [review]

File: local_simulation_server.py [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+    if not cmd[0].endswith('webots'):
+        message = f'FAIL: \'{cmd[0]}\'is not recognized as a Webots executable.'
+        close_connection(connection, message)
+        continue
+
+    world_file = cmd[-1]
```
Comment (omichel) [inline local_simulation_server.py]: We shouldn't assume the world file is at the last position either:
```suggestion
    for argument in cmd:
        if argument.endswith('.wbt')
            world_file = argument
            break
```

File: local_simulation_server.py [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+
+tcp_socket.bind((HOST, PORT))
+tcp_socket.listen()
+while True:
+    print(f'Waiting for connection on port {PORT}...')
+    connection, address = tcp_socket.accept()
+
+    print(f'Connection from {address}')
+    data = connection.recv(1024)
+    cmd = data.decode('utf-8').split(' ')
```
Comment (omichel) [inline local_simulation_server.py]: abbreviation:
```suggestion
    command= data.decode('utf-8').split(' ')
```

**Review by omichel — CHANGES_REQUESTED** [review]
Still a couple of requests...

File: local_simulation_server.py [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+
+"""Local simulation server."""
+
+import os
+import socket
+import subprocess
+import sys
+
+HOST = ''  # Any host can connect
+PORT = int(sys.argv[1])  # Port to listen on
```
Comment (omichel) [inline local_simulation_server.py]: I would set the port by default to 2000 to match the default value set in `webots_tcp_client.py`.

**Review by omichel — CHANGES_REQUESTED** [review]

File: local_simulation_server.py [hunk 27]
```diff
@@ -36,6 +36,8 @@ def close_connection(connection, message):
 tcp_socket.bind((HOST, PORT))
 tcp_socket.listen()
 while True:
+    if len(sys.argv) >= 2:
+        PORT = sys.argv[1]
```
Comment (omichel) [inline local_simulation_server.py]: Why putting this in the loop, can't we set it earlier?

File: local_simulation_server.py [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+
+"""Local simulation server."""
+
+import os
+import socket
+import subprocess
+import sys
+
+HOST = ''  # Any host can connect
+PORT = 2000  # Port to listen on
```
Comment (omichel) [inline local_simulation_server.py]:
```suggestion
PORT = 2000 if len(sys.argv) < 2 else int(sys.argv[1])  # Port to listen on
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Review by omichel — APPROVED** [review]
Looks good to me.
Thank you.

**Pull request merged**